### PR TITLE
modify to the whep protocol

### PR DESCRIPTION
**Pull request by fastfeee**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -9,3 +9,4 @@ tower-http = { version = "0.4.0", features = ["fs", "trace"] }
 webrtc = "0.8.0"
 anyhow = "1.0"
 tokio = { version = "1.15", features = ["full"] }
+hyper = "0.14"
```
Comment (a-wing) [inline Cargo.toml:12]: Where is use?

File: src/main.rs [hunk 2]
```diff
@@ -1,5 +1,14 @@
 use anyhow::Result;
-use axum::{extract::State, routing::post, Json, Router};
+use axum::{
+    //body::Body,
+    extract::{State,Request}, 
+    http::StatusCode,
+    routing::post, 
+    //Json, use axum::response::Response;
```
Comment (a-wing) [inline src/main.rs:7]: Remove this line

File: src/main.rs [hunk 3]
```diff
@@ -14,6 +23,7 @@ use webrtc::{
     peer_connection::configuration::RTCConfiguration,
     peer_connection::peer_connection_state::RTCPeerConnectionState,
     peer_connection::sdp::session_description::RTCSessionDescription,
+    //peer_connection::sdp::sdp_type::RTCSdpType,
```
Comment (a-wing) [inline src/main.rs:26]: Remove this line

File: src/main.rs [hunk 4]
```diff
@@ -182,22 +192,40 @@ async fn main() {
...
 async fn webrtc_handler(
     State(state): State<SharedState>,
-    Json(offer): Json<RTCSessionDescription>,
-) -> Json<RTCSessionDescription> {
-    let (answer, sender) = start_webrtc(offer).await;
+    request:Request,
+) -> 
+   // Result<(),(StatusCode,String)>
+   impl IntoResponse
+       {
```
Comment (a-wing) [inline src/main.rs:203]: cargo fmt

File: src/main.rs [hunk 5]
```diff
@@ -182,22 +192,40 @@ async fn main() {
...
+    //body
+    let body=request.into_body();
+    //bytes
+    let body_bytes = hyper::body::to_bytes(body).await.unwrap();
+    //String
```
Comment (a-wing) [inline src/main.rs:208]: Remove type comment, the type should auto show in your editor

For example: vscode

image


https://code.visualstudio.com/docs/languages/rust

File: src/main.rs [hunk 6]
```diff
@@ -182,22 +192,40 @@ async fn main() {
...
+    let body_string = String::from_utf8_lossy(&body_bytes).to_string();
+    //new RTCSessionDescription
+    let  whep_offer = RTCSessionDescription::offer(body_string).unwrap(); 
+    let (answer, sender) = start_webrtc(whep_offer).await;
     state.write().unwrap().ch.lock().unwrap().push(sender);
-    Json(answer)
+    (
+        StatusCode::CREATED,
+        [(header::CONTENT_TYPE,"application/sdp")],
+        [(header::LOCATION,"http://localhost:3000/whep/endpoint")],
```
Comment (a-wing) [inline src/main.rs:217]: Why have `header::LOCATION`?

**Review by a-wing — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 6 above)
Comment (fastfeee) [reply src/main.rs:217]: https://datatracker.ietf.org/doc/draft-murillo-whep/

File: src/main.rs (hunk 5 above)
Comment (fastfeee) [reply src/main.rs:208]: ok

File: src/main.rs (hunk 4 above)
Comment (fastfeee) [reply src/main.rs:203]: ok

File: src/main.rs (hunk 3 above)
Comment (fastfeee) [reply src/main.rs:26]: ok

File: src/main.rs (hunk 2 above)
Comment (fastfeee) [reply src/main.rs:7]: ok

File: Cargo.toml (hunk 1 above)
Comment (fastfeee) [reply Cargo.toml:12]: 从http请求的body转成字节用到这种方法let body_bytes = hyper::body::to_bytes(body).await.unwrap();

**Pull request merged**